src/intelligence_layer/connectors/studio/studio.py

```diff
@@ -318,6 +362,47 @@ def submit_dataset(
         self._raise_for_status(response)
         return str(response.json())
 
+    def get_dataset_examples(
+        self,
+        dataset_id: str,
+        input_type: type[Input],
+        expected_output_type: type[ExpectedOutput],
+    ) -> Iterable[StudioExample[Input, ExpectedOutput]]:
+        buffer_size = 200
+        page_size = 100
+        page: int | None = 1
+        buffer: list[StudioExample[Input, ExpectedOutput]] = []
+
+        while True:
+            if len(buffer) < buffer_size // 2 and page is not None:
+                page_url = urljoin(
+                    self.url,
+                    f"/api/projects/{self.project_id}/evaluation/datasets/{dataset_id}/datapoints?page={page}&size={page_size}",
+                )
+
+                response = requests.get(page_url, headers=self._headers)
+
+                if response.status_code == 200:
+                    examples = GetDatasetExamplesResponse(**response.json()).items
+                    buffer.extend(examples)
+
+                    if len(examples) < page_size:
+                        page = None
+                    else:
+                        page += 1
+                else:
+                    raise Exception(
+                        f"Failed to fetch items from {page_url}. Status code: {response.status_code}"
+                    )
+
+            if buffer:
+                yield StudioExample[
+                    input_type, expected_output_type  # type: ignore
+                ].model_validate_json(json_data=buffer.pop(0).model_dump_json())
```

Comment:
I think this could be really expensive memory re-allocation wise, but I am not completely sure how python handles this. It could be that the whole list has to be re-alocated if the first element is popped. It might be worth to look into this and check if maybe an ordered set or a dedicated queue data structure is better, because it has insert and delete in O(1).

Reply:
Exchanged this with a deque just to be sure